File: ontrack-extension-gitlab/src/main/java/net/nemerosa/ontrack/extension/gitlab/property/GitLabProjectConfigurationPropertyType.java
package net.nemerosa.ontrack.extension.gitlab.property;

import com.fasterxml.jackson.databind.JsonNode;
import net.nemerosa.ontrack.common.MapBuilder;
import net.nemerosa.ontrack.extension.git.property.AbstractGitProjectConfigurationPropertyType;
import net.nemerosa.ontrack.extension.gitlab.GitLabExtensionFeature;
import net.nemerosa.ontrack.extension.gitlab.model.GitLabConfiguration;
import net.nemerosa.ontrack.extension.gitlab.service.GitLabConfigurationService;
import net.nemerosa.ontrack.json.JsonUtils;
import net.nemerosa.ontrack.model.security.ProjectConfig;
import net.nemerosa.ontrack.model.security.SecurityService;
import net.nemerosa.ontrack.model.structure.ProjectEntity;
import net.nemerosa.ontrack.model.structure.ProjectEntityType;
import net.nemerosa.ontrack.model.support.ConfigurationPropertyType;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.EnumSet;
import java.util.Set;
import java.util.function.Function;

@Component
public class GitLabProjectConfigurationPropertyType
        extends AbstractGitProjectConfigurationPropertyType<GitLabProjectConfigurationProperty>
        implements ConfigurationPropertyType<GitLabConfiguration, GitLabProjectConfigurationProperty> {

    private final GitLabConfigurationService configurationService;

    @Autowired
    public GitLabProjectConfigurationPropertyType(
            GitLabExtensionFeature extensionFeature,
            GitLabConfigurationService configurationService) {
        super(extensionFeature);
        this.configurationService = configurationService;
    }

    @Override
    public String getName() {
        return "GitLab configuration";
    }

    @Override
    public String getDescription() {
        return "Associates the project with a GitLab repository";
    }

    @Override
    public Set<ProjectEntityType> getSupportedEntityTypes() {
        return EnumSet.of(ProjectEntityType.PROJECT);
    }

    @Override
    public boolean canEdit(ProjectEntity entity, SecurityService securityService) {
        return securityService.isProjectFunctionGranted(entity.projectId(), ProjectConfig.class);
    }

    @Override
    public boolean canView(ProjectEntity entity, SecurityService securityService) {
        return true;
    }

    @Override
    public GitLabProjectConfigurationProperty fromClient(JsonNode node) {
        return fromStorage(node);
    }

    @Override
    public GitLabProjectConfigurationProperty fromStorage(JsonNode node) {
        String configurationName = node.path("configuration").asText();
        // Looks the configuration up
        GitLabConfiguration configuration = configurationService.getConfiguration(configurationName);
        // OK
        return new GitLabProjectConfigurationProperty(
                configuration,
                JsonUtils.get(node, "issueServiceConfigurationIdentifier", null),
                node.path("repository").asText(),
                node.path("indexationInterval").asInt()
        );
    }

    @Override
    public JsonNode forStorage(GitLabProjectConfigurationProperty value) {
        return format(
                MapBuilder.params()
                        .with("configuration", value.getConfiguration().getName())
                        .with("repository", value.getRepository())
                        .with("indexationInterval", value.getIndexationInterval())
                        .with("issueServiceConfigurationIdentifier", value.getIssueServiceConfigurationIdentifier())
                        .get()
        );
    }

    @Override
    public GitLabProjectConfigurationProperty replaceValue(@NotNull GitLabProjectConfigurationProperty value, Function<String, String> replacementFunction) {
        return new GitLabProjectConfigurationProperty(
                configurationService.replaceConfiguration(value.getConfiguration(), replacementFunction),
                value.getIssueServiceConfigurationIdentifier(),
                replacementFunction.apply(value.getRepository()),
                value.getIndexationInterval()
        );
    }

}

File: ontrack-extension-gitlab/src/main/java/net/nemerosa/ontrack/extension/gitlab/property/GitLabProjectConfigurationPropertyType.kt
package net.nemerosa.ontrack.extension.gitlab.property

import com.fasterxml.jackson.databind.JsonNode
import net.nemerosa.ontrack.extension.git.property.AbstractGitProjectConfigurationPropertyType
import net.nemerosa.ontrack.extension.gitlab.GitLabExtensionFeature
import net.nemerosa.ontrack.extension.gitlab.model.GitLabConfiguration
import net.nemerosa.ontrack.extension.gitlab.service.GitLabConfigurationService
import net.nemerosa.ontrack.json.asJson
import net.nemerosa.ontrack.json.getTextField
import net.nemerosa.ontrack.model.security.ProjectConfig
import net.nemerosa.ontrack.model.security.SecurityService
import net.nemerosa.ontrack.model.structure.ProjectEntity
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import net.nemerosa.ontrack.model.support.ConfigurationPropertyType
import org.springframework.stereotype.Component
import java.util.*

@Component
class GitLabProjectConfigurationPropertyType(
    extensionFeature: GitLabExtensionFeature,
    private val configurationService: GitLabConfigurationService
) : AbstractGitProjectConfigurationPropertyType<GitLabProjectConfigurationProperty>(extensionFeature),
    ConfigurationPropertyType<GitLabConfiguration, GitLabProjectConfigurationProperty> {
    override val name: String
        get() = "GitLab configuration"

    override val description: String
        get() = "Associates the project with a GitLab repository"

    override val supportedEntityTypes: Set<ProjectEntityType>
        get() = EnumSet.of(ProjectEntityType.PROJECT)

    override fun canEdit(entity: ProjectEntity, securityService: SecurityService): Boolean {
        return securityService.isProjectFunctionGranted(entity.projectId(), ProjectConfig::class.java)
    }

    override fun canView(entity: ProjectEntity, securityService: SecurityService): Boolean {
        return true
    }

    override fun fromClient(node: JsonNode): GitLabProjectConfigurationProperty {
        return fromStorage(node)
    }

    override fun fromStorage(node: JsonNode): GitLabProjectConfigurationProperty {
        val configurationName = node.path("configuration").asText()
        // Looks the configuration up
        val configuration = configurationService.getConfiguration(configurationName)
        // OK
        return GitLabProjectConfigurationProperty(
            configuration = configuration,
            issueServiceConfigurationIdentifier = node.getTextField("issueServiceConfigurationIdentifier"),
            repository = node.path("repository").asText(),
            indexationInterval = node.path("indexationInterval").asInt()
        )
    }

    override fun forStorage(value: GitLabProjectConfigurationProperty): JsonNode =
        mapOf(
            "configuration" to value.configuration.name,
            "repository" to value.repository,
            "indexationInterval" to value.indexationInterval,
            "issueServiceConfigurationIdentifier" to value.issueServiceConfigurationIdentifier,
        ).asJson()

    override fun replaceValue(
        value: GitLabProjectConfigurationProperty,
        replacementFunction: (String) -> String
    ): GitLabProjectConfigurationProperty {
        return GitLabProjectConfigurationProperty(
            value.configuration,
            value.issueServiceConfigurationIdentifier,
            replacementFunction(value.repository),
            value.indexationInterval
        )
    }
}
